mitiq/benchmarks/mirror_circuits.py

```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python
+# coding: utf-8
+# %%
+
+# %%
+
+
+import cirq
+import numpy as np
+from numpy import random
+import networkx as nx
+from typing import Optional
+from cirq.experiments.qubit_characterizations import _single_qubit_cliffords
+
+
+# %%
+paulis = [cirq.X, cirq.Y, cirq.Z, cirq.I]
+
+def random_paulis(nqubits: int) -> cirq.Circuit:
+    """Returns a circuit with a random pauli gate applied to each qubit
+
+    Args:
+    nqubits: The number of qubits in the circuit
+    """
+    circuit = cirq.Circuit()
+    circuit.append((paulis[random.randint(4)](cirq.LineQubit(x)) for x in range(nqubits)))
+    return circuit
+
+
+# %%
+
+
+def edge_grab(xi: float, connectivity_graph_: nx.Graph) -> nx.Graph:
+    """Returns a set of edges for which two qubit gates are to be applied given a two qubit gate density, xi,
+    and the connectivity graph that must be satisfied
+
+    Args:
+        xi: Two qubit gate density
+        connectivity_graph_: The connectivity graph for the backend on which the circuit will be run
+    """
+    connectivity_graph = nx.Graph()
+    connectivity_graph.add_nodes_from(connectivity_graph_)
+    connectivity_graph.add_edges_from(connectivity_graph_.edges)
+    candidate_edges = nx.Graph()
+    
+    nqubits = connectivity_graph.number_of_nodes()
+
+    while(len(list(connectivity_graph.edges))>0):
+        num = random.randint(connectivity_graph.size())
+        edges = list(connectivity_graph.edges)
+        curr_edge = edges[num]
+        q1, q2 = curr_edge
+        candidate_edges.add_edge(*curr_edge)
+        connectivity_graph.remove_node(q1)
+        connectivity_graph.remove_node(q2)
+    prob = nqubits*xi / len(list(candidate_edges.edges))
+
+    final_edges = nx.Graph()
+    for edge in list(candidate_edges.edges):
+        if(random.random()<prob):
+            final_edges.add_edge(*edge)
+    return final_edges
+
+
+# %%
+
+
+single_q_cliffords = _single_qubit_cliffords()
+
+
+# %%
+
+
+def random_cliffords(nqubits: int, edges: list) -> cirq.Circuit:
+    """Returns a circuit with a two qubit clifford gate applied to each edge in edges, and a random single qubit
+    clifford gate applied to every other qubit
+
+    Args:
+        nqubits: The number of qubits in the circuit
+        edges: The edges for which the two qubit clifford gate is to be applied
+    """
+    cliffords = single_q_cliffords.c1_in_xy
+
+    gates = [cirq.CNOT(cirq.LineQubit(a), cirq.LineQubit(b)) for a,b in edges]
+
+    for x in range (nqubits):
+        notEdge = True
+        for a,b in edges:
+            if(a==x or b==x):
+                notEdge = False
+        if(notEdge):
+            num = random.randint(24)
+            for clifford_gate in cliffords[num]:
+                gates.append(clifford_gate(cirq.LineQubit(x)))
+    return cirq.Circuit(gates)
+
+
+# %%
+
+
+def random_single_cliffords(nqubits: int) -> cirq.Circuit:
```

Comment:
This function is identical to the latter half of `random_cliffords`. You should use this function in `random_cliffords` or otherwise refactor to get rid of the duplication.